src/crypto/openssl.rs

```diff
@@ -0,0 +1,941 @@
+                } else if *message_digest == MessageDigest::sha384() {
+                    SecAlg::ECDSAP384SHA384
+                } else {
+                    unreachable!();
+                };
+
+                let key = public_key.public_key();
+                let group = public_key.group();
+                let mut ctx = BigNumContext::new().expect("should not fail");
+                let key = key
+                    .to_bytes(
+                        group,
+                        PointConversionForm::UNCOMPRESSED,
+                        &mut ctx,
+                    )
+                    .expect("should not fail");
+
+                // Openssl has an extra byte with the value 4 in front.
+                let key = key[1..].to_vec();
+
+                Dnskey::new(*flags, 3, alg, key).expect("should not fail")
+            }
+        }
+    }
+}
+
+#[cfg(feature = "unstable-crypto-sign")]
+/// Submodule for private keys and signing.
+pub mod sign {
+    use std::boxed::Box;
+    use std::vec::Vec;
+
+    use crate::base::iana::SecAlg;
+    use crate::crypto::sign::{
+        GenerateParams, RsaSecretKeyBytes, SecretKeyBytes, SignError,
+        SignRaw, Signature,
+    };
+    use crate::rdata::Dnskey;
+
+    use super::{FromBytesError, GenerateError, PublicKey};
+
+    use openssl::bn::BigNum;
+    use openssl::ec::{EcGroup, EcKey};
+    use openssl::ecdsa::EcdsaSig;
+    use openssl::error::ErrorStack;
+    use openssl::hash::MessageDigest;
+    use openssl::nid::Nid;
+    use openssl::pkey::{self, Id, PKey, Private};
+    use openssl::rsa::Rsa;
+
+    use secrecy::ExposeSecret;
+
+    //----------- KeyPair ----------------------------------------------------
+
+    /// A key pair backed by OpenSSL.
+    #[derive(Clone, Debug)]
+    pub struct KeyPair {
+        /// The algorithm used by the key.
+        algorithm: SecAlg,
+
+        /// Flags from [`Dnskey`].
+        flags: u16,
+
+        /// The private key.
+        pkey: PKey<Private>,
+    }
+
+    //--- Conversion to and from bytes
+
+    impl KeyPair {
+        /// Import a key pair from bytes into OpenSSL.
+        pub fn from_bytes<Octs>(
+            secret: &SecretKeyBytes,
+            public: &Dnskey<Octs>,
+        ) -> Result<Self, FromBytesError>
+        where
+            Octs: AsRef<[u8]>,
+        {
+            /// Create a [`BigNum`] from a slice.
+            fn num(slice: &[u8]) -> Result<BigNum, FromBytesError> {
+                let mut v = BigNum::new()?;
+                v.copy_from_slice(slice)?;
+                Ok(v)
+            }
+
+            /// Create a [`BigNum`] from a slice with secure storage.
+            fn secure_num(slice: &[u8]) -> Result<BigNum, FromBytesError> {
+                let mut v = BigNum::new_secure()?;
+                v.copy_from_slice(slice)?;
+                Ok(v)
+            }
+
+            let pkey = match secret {
+                SecretKeyBytes::RsaSha256(s) => {
+                    let n = num(&s.n)?;
```

Comment:
Why not call `PublicKey::from_dnskey()` here which has this code and also would make it more consistent with the other variants handled here, and could also be used to support for RSASHA1 and RSASHA512?